File: frontend/js/views/gross_rent_multiplier.js
// Gross rent multiplier — price ÷ gross annual rent, with vacancy/credit
// loss and other income rolled into effective gross income and an effective
// GRM, via /calc/grm. Updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const FIELDS = [
    ['property_price_usd', 'Property price ($)', 600000],
    ['gross_monthly_rent_usd', 'Scheduled monthly rent ($)', 5000],
    ['vacancy_rate_pct', 'Vacancy + credit loss (%)', 5],
    ['other_income_monthly_usd', 'Other monthly income ($)', 0],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });
const VIEW = 'grm';
let lastReport = null;
let lastBody = null;

export async function renderGrm(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.grm.h1.title">// GROSS RENT MULTIPLIER</span></h1>
        <p class="muted small" data-i18n="view.grm.hint.intro">
            A fast way to compare rental prices: property price ÷ gross annual rent. A lower GRM
            is cheaper relative to the rent it produces. Scheduled rent overstates reality, so
            vacancy and credit loss are netted out (and other income added) into effective gross
            income — and an effective GRM on what's actually collected. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.grm.h2.inputs">The property</h2>
            <form id="grm-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.grm.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
            <div id="grm-tools" class="ce-toolbar"></div>
            <button type="button" id="grm-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="grm-sens" class="ce-sens"></div>
        </div>
        <div id="grm-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#grm-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        return body;
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcGrm(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.grm.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#grm-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'grm.csv' });
    mount.querySelector('#grm-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['gross_rent_multiplier', r.gross_rent_multiplier],
        ['effective_grm', r.effective_grm],
        ['effective_gross_income_usd', r.effective_gross_income_usd],
        ['gross_scheduled_income_usd', r.gross_scheduled_income_usd],
        ['vacancy_loss_usd', r.vacancy_loss_usd],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#grm-result');
    // Line chart: GRM as property price sweeps 0.5× → 1.5× current (linear in price).
    const base = body.property_price_usd || 600000;
    const xs = enh.linspace(base * 0.5, base * 1.5, 13);
    const pts = await Promise.all(xs.map(async (p) => {
        const rr = await api.calcGrm({ ...body, property_price_usd: p });
        return { x: p / 1000, y: rr ? rr.gross_rent_multiplier : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'price $k', ylabel: 'GRM' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.grm.h2.result">The multiplier</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.grm.card.grm">Gross rent multiplier</div>
                    <div class="value">${Number(r.gross_rent_multiplier).toFixed(1)}</div></div>
                <div class="card"><div class="label" data-i18n="view.grm.card.effgrm">Effective GRM</div>
                    <div class="value">${Number(r.effective_grm).toFixed(1)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.grm.card.egi">Effective gross income</div>
                    <div class="value pos">${money(r.effective_gross_income_usd)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <thead><tr><th data-i18n="view.grm.col.line">Line</th><th data-i18n="view.grm.col.amount">Annual</th></tr></thead>
                <tbody>
                    <tr><td data-i18n="view.grm.row.gsi">Gross scheduled income</td><td>${money(r.gross_scheduled_income_usd)}</td></tr>
                    <tr><td data-i18n="view.grm.row.vacancy">Vacancy + credit loss</td><td class="neg">-${money(r.vacancy_loss_usd)}</td></tr>
                    <tr><td data-i18n="view.grm.row.other">Other income</td><td class="pos">+${money(r.other_income_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.grm.row.egi">Effective gross income</td><td>${money(r.effective_gross_income_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#grm-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: monthly rent 0.5× → 1.5×; y: property price 0.5× → 1.5×. Output: effective GRM (lower is better).
    const rent = base.gross_monthly_rent_usd || 5000;
    const price = base.property_price_usd || 600000;
    const xVals = enh.linspace(rent * 0.5, rent * 1.5, 5);
    const yVals = enh.linspace(price * 0.5, price * 1.5, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'gross_monthly_rent_usd', yKey: 'property_price_usd', xVals, yVals, compute: (b) => api.calcGrm(b), pick: (r) => (r ? r.effective_grm : null) });
    if (!viewIsCurrent(tok)) return;
    // Lower GRM is a better deal, so negate for shading (green = lower) while showing the real value.
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells: cells.map((row) => row.map((v) => (v == null ? null : -v))), fmt: (v) => (v == null ? '—' : (-v).toFixed(1)), xfmt: (v) => '$' + Math.round(v), yfmt: (v) => '$' + Math.round(v / 1000) + 'k', xName: t('view.grm.label.gross_monthly_rent_usd') || 'Rent', yName: t('view.grm.label.property_price_usd') || 'Price' });
}

File: crates/traderview-core/src/gross_rent_multiplier.rs
//! Gross rent multiplier & effective gross income — rental valuation screen.
//!
//! GRM is a fast way to compare rental prices: property price ÷ gross annual
//! rent. A lower GRM is cheaper relative to the rent it produces. But
//! scheduled rent overstates reality — vacancy and credit loss take a cut,
//! while laundry/parking/fees add a bit back. Netting those gives effective
//! gross income (EGI) and a truer "effective GRM":
//!
//!   * gross scheduled income = monthly rent × 12
//!   * GRM = price / gross scheduled income
//!   * EGI = gross scheduled − vacancy/credit loss + other income
//!   * effective GRM = price / EGI
//!
//! Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct GrmInput {
    pub property_price_usd: f64,
    /// Scheduled monthly rent with all units occupied.
    pub gross_monthly_rent_usd: f64,
    /// Vacancy + credit loss as a percent of scheduled rent.
    #[serde(default)]
    pub vacancy_rate_pct: f64,
    /// Other monthly income (laundry, parking, fees).
    #[serde(default)]
    pub other_income_monthly_usd: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct GrmResult {
    pub gross_scheduled_income_usd: f64,
    pub gross_rent_multiplier: f64,
    pub vacancy_loss_usd: f64,
    pub other_income_usd: f64,
    /// Scheduled − vacancy/credit loss + other income.
    pub effective_gross_income_usd: f64,
    /// Price ÷ EGI (GRM on the income actually collected).
    pub effective_grm: f64,
}

pub fn analyze(i: &GrmInput) -> GrmResult {
    let gsi = i.gross_monthly_rent_usd.max(0.0) * 12.0;
    let grm = if gsi > 0.0 { i.property_price_usd / gsi } else { 0.0 };

    let vacancy_loss = gsi * i.vacancy_rate_pct / 100.0;
    let other_income = i.other_income_monthly_usd.max(0.0) * 12.0;
    let egi = gsi - vacancy_loss + other_income;
    let effective_grm = if egi > 0.0 { i.property_price_usd / egi } else { 0.0 };

    GrmResult {
        gross_scheduled_income_usd: gsi,
        gross_rent_multiplier: grm,
        vacancy_loss_usd: vacancy_loss,
        other_income_usd: other_income,
        effective_gross_income_usd: egi,
        effective_grm,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> GrmInput {
        GrmInput {
            property_price_usd: 600_000.0,
            gross_monthly_rent_usd: 5_000.0,
            vacancy_rate_pct: 5.0,
            other_income_monthly_usd: 0.0,
        }
    }

    #[test]
    fn grm_is_price_over_gross_annual() {
        // 600k / (5000×12 = 60k) = 10.
        let r = analyze(&base());
        assert!((r.gross_scheduled_income_usd - 60_000.0).abs() < 1e-6);
        assert!((r.gross_rent_multiplier - 10.0).abs() < 1e-9);
    }

    #[test]
    fn vacancy_loss_is_pct_of_scheduled() {
        // 5% of 60k = 3,000.
        let r = analyze(&base());
        assert!((r.vacancy_loss_usd - 3_000.0).abs() < 1e-6);
    }

    #[test]
    fn egi_nets_vacancy() {
        // 60k − 3k = 57k.
        let r = analyze(&base());
        assert!((r.effective_gross_income_usd - 57_000.0).abs() < 1e-6);
    }

    #[test]
    fn effective_grm_uses_egi() {
        // 600k / 57k ≈ 10.526.
        let r = analyze(&base());
        assert!((r.effective_grm - 600_000.0 / 57_000.0).abs() < 1e-6);
        assert!(r.effective_grm > r.gross_rent_multiplier); // EGI < scheduled
    }

    #[test]
    fn other_income_raises_egi() {
        // +200/mo = 2,400/yr → EGI 57k + 2.4k = 59.4k.
        let r = analyze(&GrmInput { other_income_monthly_usd: 200.0, ..base() });
        assert!((r.other_income_usd - 2_400.0).abs() < 1e-6);
        assert!((r.effective_gross_income_usd - 59_400.0).abs() < 1e-6);
    }

    #[test]
    fn lower_price_lower_grm() {
        let cheap = analyze(&GrmInput { property_price_usd: 480_000.0, ..base() });
        let dear = analyze(&base());
        assert!(cheap.gross_rent_multiplier < dear.gross_rent_multiplier);
    }

    #[test]
    fn no_vacancy_egi_equals_scheduled() {
        let r = analyze(&GrmInput { vacancy_rate_pct: 0.0, ..base() });
        assert!((r.effective_gross_income_usd - r.gross_scheduled_income_usd).abs() < 1e-6);
        assert!((r.effective_grm - r.gross_rent_multiplier).abs() < 1e-9);
    }

    #[test]
    fn zero_rent_guards() {
        let r = analyze(&GrmInput { gross_monthly_rent_usd: 0.0, ..base() });
        assert!(r.gross_rent_multiplier.abs() < 1e-9);
        assert!(r.effective_grm.abs() < 1e-9);
    }
}
